Improving S3 documentation.

elsonrodriguez opened the pull request:
Added a copy-pastable guide on the variables, and also provided usable examples that give immediate feedback.

elsonrodriguez committed:
Improving S3 documentation.

Added a copy-pastable guide on the variables, and also provided usable examples that give immediate feedback.

yongtang commented on tensorflow/docs_src/deploy/s3.md:
````diff
@@ -1,12 +1,57 @@
+
+```bash
+AWS_ACCESS_KEY_ID=XXXXX                 # Credentials only needed if connecting to a private endpoint
+AWS_SECRET_ACCESS_KEY=XXXXX
````
I think `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` is not a must as you could allow credentials to be setup through other means. In fact, it might not be desired as it is visible in env.

elsonrodriguez replied:
I've made a note of this below, but wanted people to be able to copy/paste to get something working quickly.

I can expand the description here to include a warning about secrets as env vars, and also mention the improved security of the other methods.

yongtang replied:
Thanks @elsonrodriguez. I am thinking we keep the original descriptions of environmental variables and setting credential alone. Instead, we could create another section of example at the beginning. In the example section we could copy the setup you added (with ENV for `AWS_ACCESS_KEY_ID`/etc) and the example you added.

elsonrodriguez replied:
I've added back the section, however it feels redundant, should I get rid of the variable comment descriptions in the setup section?

yongtang replied:
Thanks @elsonrodriguez. As most of the ENVs are not necessary (e.g., `S3_USE_HTTPS` will be used unless `S3_USE_HTTPS=0` is provided), maybe we could remove most of the ENVs here.

We could keep the ENVs of `AWS_ACCESS_KEY_ID=XXXXX` and `AWS_SECRET_ACCESS_KEY=XXXXX` in the example. As the user will need to provide credentials one way (ENV) or another (profiles, etc).

elsonrodriguez replied:
If you put it that way, I think maybe it's fine as is.

Most of my testing lately has been with Minio, which I deploy without SSL/HTTPS. If I were quickly skimming this example I might miss that setting and be confused.

I think I'm ok with the redundancy given the clarity in both the example and configuration guide.

yongtang replied:
Thanks @elsonrodriguez. If the current documentation format helps I think we can leave it as it. 
Though the usage of word "must" might be strong. Can you change the sentence to:
`To configure your Tensorflow applications to communicate to an S3 endpoint, you could set the following environment variables:`

Also, is it possible to move the `Configuration Notes` section before the `Setup` section? I am concerned that asking users to provide `ENV` explicitly at the beginning of the doc may mislead and make users to believe that `ENVs` are necessary. It would be better served as only an "example setup" I think.

elsonrodriguez replied:
That makes sense, I rearranged the doc and change the sentence to be softer and more contextual.

yongtang commented on tensorflow/docs_src/deploy/s3.md:
```diff
@@ -1,12 +1,57 @@
+* [Amazon S3](https://aws.amazon.com/s3/)
+* [Google Storage](https://cloud.google.com/storage/docs/interoperability)
+* [Minio](https://www.minio.io/kubernetes.html)
+* [Ceph](http://docs.ceph.com/docs/master/radosgw/)
```
S3 was tested against AWS before. It is interesting to see there are other implementations. Maybe the tests could be run against those implementations?

elsonrodriguez replied:
I've manually tested against AWS, Google, and Minio (standalone, distributed has a bug). I haven't tested Ceph with TF but I'm told Radosgw does the job, and felt fair to represent it.

Or are you talking about automated tests?

yongtang replied:
Thanks @elsonrodriguez. if S3 works on those systems I think we could keep it. Would you mind move these references to the end of the page? That way user will be easier to find the solution they need without reading through.

elsonrodriguez replied:
Fixed, also removed Ceph since I didn't test it.

yongtang replied:
Thanks for the update 👍 !

elsonrodriguez committed:
Updating docs based on feedback.

Added back old configuration section, moved s3 implementations to bottom of document.

elsonrodriguez committed:
Rearranged documentation before example, renamed sections to be more clear.

yongtang reviewed (commented):
Thanks @elsonrodriguez, LGTM!
